### PR TITLE
Add hmpps-contacts-dev namespace

### PR DESCRIPTION
**Pull request by prabash-moj**
initial commit to create project templates and namespaces for hmpps contacts project

### CODE REVIEW

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/00-namespace.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    cloud-platform.justice.gov.uk/is-production: "false"
+    cloud-platform.justice.gov.uk/environment-name: "development"
+    pod-security.kubernetes.io/enforce: restricted
+  annotations:
+    cloud-platform.justice.gov.uk/business-unit: "HMPPS"
+    cloud-platform.justice.gov.uk/slack-channel: "public_move-and-improve"
+    cloud-platform.justice.gov.uk/application: "HMPPS contacts"
+    cloud-platform.justice.gov.uk/owner: "Move and Improve: move-and-improve@digital.justice.gov.uk"
+    cloud-platform.justice.gov.uk/source-code: "https://github.com/ministryofjustice/hmpps-contacts-dev-api"
+    cloud-platform.justice.gov.uk/team-name: "farsight-devs"
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/00-namespace.yaml:15]: `hmpps-contacts-api` (no environment tag) would be the name.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/01-rbac.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+metadata:
+  name: hmpps-contacts-dev-admin
+  namespace: hmpps-contacts-dev
+subjects:
+  - kind: Group
+    name: "github:dps-tech"
+    apiGroup: rbac.authorization.k8s.io
+  - kind: Group
+    name: "github:hmpps-move-and-improve"
+    apiGroup: rbac.authorization.k8s.io
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/01-rbac.yaml:12]: Are we all in GH group hmpps-move-and-improve?  Actually I'm not, but I can organise this.

Comment (timharrison-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/01-rbac.yaml:12]: You and Rosario are.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/06-certificate.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+---
+apiVersion: cert-manager.io/v1
+kind: Certificate
+metadata:
+  name: hmpps-contacts-dev-api-dev-cert
+  namespace: hmpps-contacts-dev
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/06-certificate.yaml:6]: `hmpps-contacts-api-dev-cert` - too many `dev` labels.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/06-certificate.yaml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+spec:
+  secretName: hmpps-contacts-dev-api-cert
+  issuerRef:
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/06-certificate.yaml:9]: Secrets don't need the env in the name `hmpps-contacts-api-cert` is enough.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/06-certificate.yaml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    name: letsencrypt-production
+    kind: ClusterIssuer
+  dnsNames:
+    - contacts-api-dev.hmpps.service.justice.gov.uk
+---
+apiVersion: cert-manager.io/v1
+kind: Certificate
+metadata:
+  name: hmpps-contacts-dev-ui-dev-cert
+  namespace: hmpps-contacts-dev
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/06-certificate.yaml:19]: Same for UI - don't need environment in the names.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/analytical-platform-irsa.tf [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+  # EKS configuration
+  eks_cluster_name = var.eks_cluster_name
+
+  # IRSA configuration
+  service_account_name = "hmpps-contacts-to-ap-s3"
+  namespace            = var.namespace
+  role_policy_arns = {
+    s3 = aws_iam_policy.ap_policy.arn
+  }
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/analytical-platform-irsa.tf:12]: I would remove this file from the initial build - we haven't talked to the AP team (as far as I know) so will have no jobs to export data to their S3 bucket.  We can always add this later.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/domain-events-topic.tf [hunk 7]
```diff
@@ -0,0 +1,14 @@
+resource "kubernetes_secret" "hmpps_contacts_hmpps_domain_events_topic" {
+  metadata {
+    name      = "hmpps-domain-events-topic"
+    namespace = var.namespace
+  }
+
+  data = {
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/domain-events-topic.tf:7]: Remove this from initial build too - we can integrate with domain events topic later.

Comment (timharrison-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/domain-events-topic.tf:7]: On second thoughts - please leave it in - saves us a job later.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/irsa.tf [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+module "irsa" {
+  source = "github.com/ministryofjustice/cloud-platform-terraform-irsa?ref=2.0.0" # use the latest release
+
+  # EKS configuration
+  eks_cluster_name = var.eks_cluster_name
+
+  # IRSA configuration
+  service_account_name = "hmpps-contacts-dev-api"
+  role_policy_arns     = merge(local.sns_policies)
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/irsa.tf:21]: Remove `-dev` from name - should be `hmpps-contacts-api` as service account name in all environments.

Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/irsa.tf:21]: Remove `-dev` from name - should be `hmpps-contacts-api` as service account name in all environments.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/main.tf [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  region = "eu-west-2"
+
+  default_tags {
+    tags = {
+      source-code   = "github.com/ministryofjustice/cloud-platform-environments"
+      slack-channel = var.slack_channel
+      GithubTeam = "hmpps-move-and-improve"
+    }
+  }
+}
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/main.tf:16]: This this section is additional to requirements? So london and ireland required - eu-west-1 and eu-west-2 - we have 2 x eu-west-1 definitions.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/main.tf [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+provider "github" {
+  token = var.github_token
+  owner = var.github_owner
+}
+
+provider "kubernetes" {}
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/main.tf:49]: Think we might need a section on the end here containing `locals {
  default_tags = {
    business_unit          = var.business_unit
    application            = var.application
    is_production          = var.is_production
    team_name              = var.team_name
    environment_name       = var.environment
    infrastructure_support = var.infrastructure_support
  }
}`

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/rds-postgresql.tf [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    database_password     = module.rds.database_password
+    rds_instance_address  = module.rds.rds_instance_address
+  }
+  /* You can replace all of the above with the following, if you prefer to
+     * use a single database URL value in your application code:
+     *
+     * url = "postgres://${module.rds.database_username}:${module.rds.database_password}@${module.rds.rds_instance_endpoint}/${module.rds.database_name}"
+     *
+     */
+}
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/rds-postgresql.tf:56]: Could you add this section on the end please?
It records some useful info too - 

```
resource "kubernetes_config_map" "rds" {
  metadata {
    name      = "rds-postgresql-instance-output"
    namespace = var.namespace
  }

  data = {
    database_name = module.rds.database_name
    db_identifier = module.rds.db_identifier
  }
}
```

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/serviceaccount-circleci.tf [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+module "serviceaccount" {
+  source               = "github.com/ministryofjustice/cloud-platform-terraform-serviceaccount?ref=1.1.0"
+  namespace            = var.namespace
+  kubernetes_cluster   = var.kubernetes_cluster
+  serviceaccount_name  = "circleci"
+  role_name            = "circleci"
+  serviceaccount_rules = local.sa_rules
+}
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/serviceaccount-circleci.tf:72]: May be missing a value? `rolebinding_name = "circleci"` - I'd need to check what it does though.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/variables.tf [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+variable "kubernetes_cluster" {
+  description = "Kubernetes cluster name for references to secrets for service accounts"
+  type        = string
+}
+
+variable "application" {
+  description = "Name of the application you are deploying"
+  type        = string
+  default     = "HMPPS Challenge Support Intervention Plan"
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/variables.tf:14]: Wrong app name

**Review by timharrison-moj — COMMENTED** [review]
Good first attempt - a few observations to take a look at.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/01-rbac.yaml (hunk 2 above)
Comment (prabash-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/01-rbac.yaml:12]: yeah you are not in this group yet, https://github.com/orgs/ministryofjustice/teams/hmpps-move-and-improve?query= please get your git account added to this group

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/irsa.tf (hunk 8 above)
Comment (prabash-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/irsa.tf:21]: addressed this already , looks like a duplicate comment , same as previous one, :)

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/main.tf (hunk 9 above)
Comment (prabash-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/main.tf:16]: Oh, looks like its duplicated in the 'hmpps-challenge-support-intervention-plan-dev' folder where I have copied it from , removed now form this file.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/main.tf (hunk 10 above)
Comment (prabash-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/main.tf:49]: not sure what the intention of that section, havent seen it on any other main.tf files in the other packages, but yeah added that,

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/rds-postgresql.tf (hunk 11 above)
Comment (prabash-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/rds-postgresql.tf:56]: added,

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/serviceaccount-circleci.tf (hunk 12 above)
Comment (prabash-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/serviceaccount-circleci.tf:72]: it looks like only added to prison-api and strenghths-based-needs packages. but yeah added this, we can remove it if not needed,

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/rds-postgresql.tf [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+  data = {
+    rds_instance_endpoint = module.rds.rds_instance_endpoint
+    database_name         = module.rds.database_name
+    database_username     = module.rds.database_username
+    database_password     = module.rds.database_password
+    rds_instance_address  = module.rds.rds_instance_address
+  }
+
+  resource "kubernetes_config_map" "rds" {
+    metadata {
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/rds-postgresql.tf:52]: Identation is wrong here - failing tflint too - the configmap is a separate section entirely, and not part of the secret.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/serviceaccount-circleci.tf (hunk 12 above)
Comment (timharrison-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/serviceaccount-circleci.tf:73]: Alignment of = ?

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/versions.tf [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+      version = "~> 5.39.0"
+    }
+    kubernetes = {
+      source  = "hashicorp/kubernetes"
+      version = "~> 2.23.0"
+    }
+    random = {
+      source  = "hashicorp/random"
+      version = "~> 3.6.0"
+    }
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/versions.tf:19]: Maybe lose the random provider if not used?

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/analytical-platform-irsa.tf (hunk 6 above)
Comment (prabash-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/analytical-platform-irsa.tf]: removed this file

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/versions.tf (hunk 15 above)
Comment (prabash-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/versions.tf:19]: not sure if its not used, by looking at the terraform/random provider wiki says 'Supports the use of randomness within Terraform configurations. This is a logical provider, which means that it works entirely within Terraform logic, and does not interact with any other services.' so I guess its ok to leave it for now, most of the other configurations have this too

Comment (timharrison-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/versions.tf:19]: It is still failing the tflint because of this.

Comment (timharrison-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/versions.tf:19]: Warning: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/versions.tf:16:5: Warning - provider 'random' is declared in required_providers but not used by the module (terraform_unused_required_providers)
Error: Process completed with exit code 2.

Comment (prabash-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/versions.tf]: ahha, :) removed that now

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/variables.tf (hunk 13 above)
Comment (timharrison-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/variables.tf]: This is still showing the CSIP name.

Comment (prabash-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/variables.tf]: image
have updated and pushed changes , no local changes to push , wonder if we are looking at the latest version here on github?

Comment (prabash-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/variables.tf]: it is updated here, https://github.com/ministryofjustice/cloud-platform-environments/pull/24710/files#diff-9c5b78f3c5764f34b40e66116e374643eeeb9d2f6cbad597febb057292fa9b05

Comment (timharrison-moj) [reply namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/variables.tf]: Yes - can see it now thanks. Maybe looking at an earlier commit.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/variables.tf [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+  description = "Area of the MOJ responsible for this service"
+  type        = string
+  default     = "HMPPS"
+}
+
+variable "team_name" {
+  description = "Name of the development team responsible for this service"
+  type        = string
+  default     = "farsight-devs"
+}
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/variables.tf:33]: Should this be "hmpps-move-and-improve"?

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/irsa.tf [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+  team_name              = var.team_name
+  namespace              = var.namespace # this is also used to attach your service account to your namespace
+  environment_name       = var.environment
+  infrastructure_support = var.infrastructure_support
+}
+
+data "aws_ssm_parameter" "irsa_policy_arns_sns" {
+  for_each = local.sns_topics
+  name     = "/${each.value}/sns/${each.key}/irsa-policy-arn"
+}
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-dev/resources/irsa.tf:36]: Could you add this bit at the end please? It records the IAM role service account details in a secret in the namespace: ```esource "kubernetes_secret" "irsa" {
  metadata {
    name      = "irsa-output"
    namespace = var.namespace
  }
  data = {
    role           = module.irsa.role_name
    serviceaccount = module.irsa.service_account.name
    rolearn        = module.irsa.role_arn
  }
}```

**Review by timharrison-moj — COMMENTED** [review]
Couple of minors - then good to go.

**Review by sj-williams — APPROVED** [review]

**Pull request merged**